File: 31_StackPushPopOrder/StackPushPopOrder.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <cstring>
using namespace std;

class Solution
{
  public:
    bool IsPopOrder(vector<int> pushV, vector<int> popV)
    {
        stack<int> st;
        int id = 0;
        for (int i = 0; i < popV.size(); i++)
        {
            while (st.empty() || st.top() != popV[i])
            {
                st.push(pushV[id++]);
                if (id > pushV.size())
                    return false;
            }
            st.pop();
        }
        if (st.empty())
            return true;
        else
            return false;
    }
};

int main()
{
    Solution s = Solution();
    vector<int> p{1, 2, 3, 4, 5};
    vector<int> q{4, 3, 5, 1, 2};
    bool flag = s.IsPopOrder(p, q);
    return 0;
}

File: 31_StackPushPopOrder/StackPushPopOrder.py
class Solution:
    def IsPopOrder(self, pushV, popV):
        # write code here
        stack = []
        id = 0
        for i in popV:
            while not stack or stack[-1] != i:
                if id == len(pushV):
                    break
                stack.append(pushV[id])
                id += 1
            if stack[-1] != i:
                break
            stack.pop()
        if not stack:
            return True
        else:
            return False


s = Solution()
print(s.IsPopOrder([1, 2, 3, 4, 5], [4, 3, 5, 1, 2]))
